### PR TITLE
Perf bench test

### PR DESCRIPTION
**Pull request by NiuJ1ao**

### CODE REVIEW

File: rps_bench_profile.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+DELAY=140900
+RPS=50
+else
+echo -e "VM_NUM argument invalid\n"
+exit
+fi
+
+sudo mkdir -m777 -p /tmp/ctrd-logs && sudo env "PATH=$PATH" /usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml 1>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.out 2>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.err &
+
+perf stat -D $DELAY -a -e instructions,LLC-loads,LLC-load-misses,LLC-stores,LLC-store-misses --output perf-${VM_NUM}VMs-lr_training.profile sudo env "PATH=$PATH" go test -v -race -cover -run TestBenchRequestPerSecond -args -vm $VM_NUM -requestPerSec $RPS -executionTime 2
```
Comment (ustiugov) [inline rps_bench_profile.sh:20]: `-race -cover` flags should be removed. the first one slows done a go program because it uses extra instrumentation for data race detection.

**Review by ustiugov — CHANGES_REQUESTED** [review]
please apply my comments from the previous review

File: rps_bench_profile.sh (hunk 1 above)
Comment (NiuJ1ao) [reply rps_bench_profile.sh:20]: Done

**Comment by NiuJ1ao** [comment]
Please ignore the files in the profile folder. I will delete them later.

File: Makefile [hunk 2]
```diff
@@ -104,6 +104,10 @@ bench:
 	sudo env "PATH=$(PATH)" go test $(EXTRAGOARGS) -run TestBenchRequestPerSecond -args $(WITHSNAPSHOTS)
 	./scripts/clean_fcctr.sh
 
+	sudo mkdir -m777 -p $(CTRDLOGDIR) && sudo env "PATH=$(PATH)" /usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml 1>$(CTRDLOGDIR)/fccd_orch_noupf_log_bench.out 2>$(CTRDLOGDIR)/fccd_orch_noupf_log_bench.err &
+	sudo env "PATH=$(PATH)" go test $(EXTRAGOARGS) -run TestBenchRequestPerSecond -args -benchDirTest configBase -metricsTest
```
Comment (ustiugov) [inline Makefile:108]: what is the reason for this change?

File: perf.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	args    string
+	sep     string
+	outFile string
+	run     chan bool
+	stop    chan bool
+}
+
+// NewPerfStat returns a new instance for perf stat
+func NewPerfStat(args ...string) *PerfStat {
+	perfStat := new(PerfStat)
```
Comment (ustiugov) [inline perf.go:33]: Golang has native support for runtime arguments. Please rewrite.

File: perf.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		}
+	}
+
+	if perfStat.sep == "" {
+		perfStat.sep = "|"
+		perfStat.args += delimiter + SEP + perfStat.sep
+	}
+
+	go func() {
+		for {
```
Comment (ustiugov) [inline perf.go:58]: I don't understand why you need this.

File: perf_bench_test.go [hunk 5]
```diff
@@ -37,114 +41,192 @@ import (
...
 	isColdStart     = flag.Bool("coldStart", false, "Profile cold starts (default is false)")
 	vmNum           = flag.Int("vm", 2, "The number of VMs")
-	targetReqPerSec = flag.Int("requestPerSec", 2, "The target number of requests per second")
-	executionTime   = flag.Int("executionTime", 2, "The execution time of the benchmark in seconds")
+	targetReqPerSec = flag.Int("requestPerSec", 10, "The target number of requests per second")
+	executionTime   = flag.Int("executionTime", 1, "The execution time of the benchmark in seconds")
+	funcNames       = flag.String("funcNames", "helloworld", "Name of the functions to benchmark")
+	perfEvents      = flag.String("perfEvents", "", "Perf events (run `perf stat` if not empty)")
+
+	envPath = "PATH=" + os.Getenv("PATH")
```
Comment (ustiugov) [inline perf_bench_test.go:49]: this is definitely wrong. you should not require this.

File: perf_bench_test.go [hunk 6]
```diff
@@ -37,114 +41,192 @@ import (
...
+)
+
+const (
+	AveExecTime = "AveExecTime"
```
Comment (ustiugov) [inline perf_bench_test.go:53]: naming: Ave -> Avg

File: perf_bench_test.go [hunk 7]
```diff
@@ -37,114 +41,192 @@ import (
...
+		require.True(t, testResult)
+
+		err = teardownBenchRPS()
+		require.NoError(t, err, "Teardown TestBenchRequestPerSecond error")
+	}
+}
+
+func setupBenchRPS() error {
+	// sudo mkdir -m777 -p /tmp/ctrd-logs && sudo env "PATH=$PATH" /usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml 1>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.out 2>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.err &
+	var CTRDLOGDIR = "/tmp/ctrd-logs"
```
Comment (ustiugov) [inline perf_bench_test.go:82]: a variable should not be named like that. Also, use const type for constants.

File: perf_bench_test.go [hunk 8]
```diff
@@ -37,114 +41,192 @@ import (
...
+		return err
+	}
+
+	commandString := "sudo env " + envPath + " /usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml 1>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.out 2>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.err &"
+	cmd := exec.Command("sudo", "/bin/sh", "-c", commandString)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stdout
+
+	if err := cmd.Run(); err != nil {
+		return err
```
Comment (ustiugov) [inline perf_bench_test.go:94]: that's insufficient

File: perf_bench_test.go [hunk 9]
```diff
@@ -37,114 +41,192 @@ import (
...
+	// sudo mkdir -m777 -p /tmp/ctrd-logs && sudo env "PATH=$PATH" /usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml 1>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.out 2>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.err &
+	var CTRDLOGDIR = "/tmp/ctrd-logs"
+
+	if err := os.MkdirAll(CTRDLOGDIR, 0777); err != nil {
+		return err
+	}
+
+	commandString := "sudo env " + envPath + " /usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml 1>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.out 2>/tmp/ctrd-logs/fccd_orch_noupf_log_bench.err &"
+	cmd := exec.Command("sudo", "/bin/sh", "-c", commandString)
+	cmd.Stdout = os.Stdout
```
Comment (ustiugov) [inline perf_bench_test.go:90]: why do you need this? You should use go linter, it would notify regarding the variables that are not used

File: profile/perf_profile.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+	"github.com/manifoldco/promptui"
+)
+
+var (
+	envPath = "PATH=" + os.Getenv("PATH")
+)
+
+// PerfProfileRequestsPerSeondBench profiles requests per second benchmark
+func PerfProfileRequestsPerSeondBench() error {
```
Comment (ustiugov) [inline profile/perf_profile.go:16]: typo in the name

File: profile/perf_profile.go [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+		if result != "DONE" {
+			selectedEvents = append(selectedEvents, result)
+		}
+		choice = result
+	}
+
+	return selectedEvents, nil
+}
+
+func preCommands() error {
```
Comment (ustiugov) [inline profile/perf_profile.go:111]: bad name

File: profile/perf_profile.go [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+	cmd.Stderr = os.Stdout
+
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func postCommands() error {
```
Comment (ustiugov) [inline profile/perf_profile.go:152]: bad name

File: profile/perf_profile.go [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+	return map[string]string{
+		// "helloworld":   "ustiugov/helloworld:var_workload",
+		// "chameleon":    "ustiugov/chameleon:var_workload",
+		// "pyaes":        "ustiugov/pyaes:var_workload",
+		// "image_rotate": "ustiugov/image_rotate:var_workload",
+		// "json_serdes":  "ustiugov/json_serdes:var_workload",
+		// "lr_serving":   "ustiugov/lr_serving:var_workload",
+		// "cnn_serving":  "ustiugov/cnn_serving:var_workload",
+		// "rnn_serving":  "ustiugov/rnn_serving:var_workload",
+		"lr_training": "ustiugov/lr_training:var_workload",
```
Comment (ustiugov) [inline profile/perf_profile.go:175]: You should not use training.

**Review by ustiugov — CHANGES_REQUESTED** [review]
Please write a clear description of the changes, make sure that the functionality works, tests and go linter pass (tests should pass locally for now) when asking for a review.

I am not sure that the code is working. The code quality is quite low, please be more attentive. Finally, don't forget to add unit tests.

Also, I like the idea of putting the bulk of the functionality in a separate module `profile/*`. Please continue doing so so that perf.go can be moved there as well.

File: Makefile (hunk 2 above)
Comment (NiuJ1ao) [reply Makefile:108]: I forgot to remove it after merge. Done.

File: profile/perf_profile.go (hunk 10 above)
Comment (NiuJ1ao) [reply profile/perf_profile.go:16]: perf_profile.go is deleted and its functionality has been included in perf.go

File: perf_bench_test.go (hunk 6 above)
Comment (NiuJ1ao) [reply perf_bench_test.go:53]: Done

File: perf_bench_test.go (hunk 5 above)
Comment (NiuJ1ao) [reply perf_bench_test.go:49]: Done

File: perf.go (hunk 4 above)
Comment (NiuJ1ao) [reply perf.go:58]: Have changed the implementation of invoking perf.

File: perf_bench_test.go (hunk 7 above)
Comment (NiuJ1ao) [reply perf_bench_test.go:82]: Done

File: perf.go (hunk 3 above)
Comment (NiuJ1ao) [reply perf.go:33]: The idea is to create a new instance of perf command with a specific set of arguments rather than calling this function at run time.

File: perf_bench_test.go (hunk 8 above)
Comment (NiuJ1ao) [reply perf_bench_test.go:94]: Done

File: perf_bench_test.go (hunk 9 above)
Comment (NiuJ1ao) [reply perf_bench_test.go:90]: Done

**Comment by NiuJ1ao** [comment]
The main function is now in [profile/test/perf_profile.go](https://github.com/NiuJ1ao/vhive-orchestrator/pull/1/commits/a3471983095878091089ac5d2927f7f75f881162#diff-9dbbe240041ee572014ee13c1865e074005117b5000d55e0754de1b935fc8996). With given function name and perf events, it runs [TestBenchMultiVMRequestPerSecond](https://github.com/NiuJ1ao/vhive-orchestrator/pull/1/commits/a3471983095878091089ac5d2927f7f75f881162#diff-53cfd97aedde55286611366c99402143ea83bb0d8fb6e7327d9ebc3263b78e56) for different VM numbers (4 and 8 for now) iteratively. After each iteration, it saves results in `benchRPS.csv`. When the execution finishes, it saves line charts of attributes as `<attribute_name>.png`.

File: profile/gperf.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	p.proc = pid
+
+	time.Sleep(100 * time.Millisecond) // Wait for perf completely start
+	return nil
+}
+
+// Stop stops perf stat command and returns the result in map[event]value
+func (p *PerfStat) Stop() (map[string]float64, error) {
```
Comment (ustiugov) [inline profile/gperf.go:89]: you should not kill perf. You should run it for a user-specified amount of time during for a user-specified number of rounds.

File: profile/gperf.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Common flags of perf
+const (
```
Comment (ustiugov) [inline profile/gperf.go:16]: I don't understand why we need them here.

**Review by ustiugov — CHANGES_REQUESTED** [review]
The code quality is much better. Pay more attention to naming functions, variables and files. Add a description in a comment in the main file of your module. 

Most importantly, need to change the execution model: the user runs the test and the test runs perf (just profile with `-a` system-wide) for a certain amount of time and a certain number of rounds (e.g., for 1 second 5 times). 

You don't need to run the cleaning script. It is only necessary when dealing with the snapshots due to a problem with containerd (i.e., VM shutdown executes with an error). If you need to, it is enough to shut down the VMs.

**Comment by ustiugov** [comment]
Also, you need to add unit and integration tests.

File: perf_bench_test.go [hunk 16]
```diff
@@ -47,104 +57,144 @@ func TestBenchRequestPerSecond(t *testing.T) {
...
-		// For future use
-		// createSnapshots(t, concurrency, vmID, imageName, isSyncOffload)
-		// log.Info("Snapshot created")
-		// createRecords(t, concurrency, vmID, imageName, isSyncOffload)
-		// log.Info("Record done")
-
-		funcs = append(funcs, funcName)
 	}
 
+	imagesLen := len(images)
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply perf_bench_test.go:81]: Done

File: perf_bench_test.go [hunk 17]
```diff
@@ -47,104 +57,144 @@ func TestBenchRequestPerSecond(t *testing.T) {
 		pinnedFuncNum int
 		vmID          int
 		isSyncOffload bool = true
-		images             = getImages()
-		funcs              = []string{}
+		isPerf             = len(*perfEvents) > 0
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply perf_bench_test.go:60]: Done

File: profile/test/perf_profile.go [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+func runTestBenchRequestPerSecond(vmNum int) {
+	envPath := "PATH=" + os.Getenv("PATH")
+	commandString := fmt.Sprintf("sudo env %s go test -v -run TestBenchRequestPerSecond -args -vm %d -funcNames %s -perfEvents %s",
+		envPath,
+		vmNum,
+		*funcNames,
+		*perfEvents)
+	cmd := exec.Command("sudo", "/bin/sh", "-c", commandString)
+	cmd.Stdout = os.Stdout
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/test/perf_profile.go:66]: Done

File: profile/test/perf_profile.go [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+			value, err := strconv.ParseFloat(records[row][col], 64)
+			if err != nil {
+				log.Fatalf("Failed parsing string to float: %v", err)
+			}
+			pts[row-1].Y = value
+			vmNum += 4
+		}
+
+		// plot
+		err = plotutil.AddLinePoints(p, pts)
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/test/perf_profile.go:139]: Done

File: profile/test/perf_profile.go [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+		if err != nil {
+			log.Fatalf("Failed plotting data: %v", err)
+		}
+
+		// save plot
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/test/perf_profile.go:144]: Done

File: perf_bench_test.go [hunk 21]
```diff
@@ -47,104 +57,144 @@ func TestBenchRequestPerSecond(t *testing.T) {
...
-		// "lr_training":  "ustiugov/lr_training:var_workload",
+func writeResultToCSV(t *testing.T, metrics map[string]float64, filePath string) {
+	f, err := os.OpenFile(filePath, os.O_RDWR|os.O_APPEND, 0644)
+	require.NoError(t, err, "Failed opening file")
+	defer f.Close()
+
+	reader := csv.NewReader(f)
+	records, err := reader.ReadAll()
+	if err != nil {
+		log.Fatalf("Failed reading file: %v", err)
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply perf_bench_test.go:181]: Done

File: perf_bench_test.go [hunk 22]
```diff
@@ -47,104 +57,144 @@ func TestBenchRequestPerSecond(t *testing.T) {
...
+func getImages(t *testing.T) []string {
+	var (
+		images = getAllImages()
+		funcs  = strings.Split(*funcNames, ",")
+		result []string
+	)
+
+	for _, funcName := range funcs {
+		imageName, isPresent := images[funcName]
+		require.True(t, isPresent, "Function is not supported")
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply perf_bench_test.go:166]: Done

File: profile/gperf.go [hunk 23]
```diff
@@ -0,0 +1,132 @@
+package profile
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/gperf.go:1]: `PerfStat.Stop` may accidentally interrupt the program if I use the name `perf.go` in my previous implementation since it kills the process by its name. Now, we don't need to kill `perf` and I can change it back. g is for go.

File: perf_bench_test.go [hunk 24]
```diff
@@ -47,104 +57,144 @@ func TestBenchRequestPerSecond(t *testing.T) {
...
-	done := make(chan bool, 1)
+	ticker := time.NewTicker(timeInterval)
+	tickerDone := make(chan bool, 1)
+
+	serveMetrics := make(map[string]float64)
+	serveMetrics[avgExecTime] = 0
+	serveMetrics[realRPS] = 0
 
-	for totalRequests > 0 {
+	remainRequests := totalRequests
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply perf_bench_test.go:104]: Done

File: perf_bench_test.go [hunk 25]
```diff
@@ -47,104 +57,144 @@ func TestBenchRequestPerSecond(t *testing.T) {
...
-		require.NoError(t, err, "Function returned error, "+message)
-	} else {
-		require.Equal(t, resp.IsColdStart, false)
+	require.Equal(t, resp.IsColdStart, false)
+
+	execTime := time.Since(tStart).Milliseconds()
+	serveMetrics[avgExecTime] += float64(execTime)
+	log.Debugf("VM %s: returned in %d milliseconds", vmIDString, execTime)
+
+	if resp.Payload == "Hello, replay_response!" {
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply perf_bench_test.go:151]: Done

File: profile/test/perf_profile.go [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+func runTestBenchRequestPerSecond(vmNum int) {
+	envPath := "PATH=" + os.Getenv("PATH")
+	commandString := fmt.Sprintf("sudo env %s go test -v -run TestBenchRequestPerSecond -args -vm %d -funcNames %s -perfEvents %s",
+		envPath,
+		vmNum,
+		*funcNames,
+		*perfEvents)
+	cmd := exec.Command("sudo", "/bin/sh", "-c", commandString)
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/test/perf_profile.go:65]: Done

File: profile/test/perf_profile.go [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+	cmd.Stderr = os.Stdout
+
+	if err := cmd.Run(); err != nil {
+		log.Fatalf("Benchmark execution error: %v\n", err)
+	}
+}
+
+func teardownBenchRPS() {
+	// ./scripts/clean_fcctr.sh
+	cmd := exec.Command("sudo", "/bin/sh", "-c", "./scripts/clean_fcctr.sh")
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/test/perf_profile.go:76]: Done

File: profile/test/perf_profile.go [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	// remove perf tmp file
+	if err := os.Remove("perf-tmp.data"); err != nil {
+		log.Warnf("Remove file error: %v", err)
+	}
+}
+
+func createResultCSV() {
+	f, err := os.Create("benchRPS.csv")
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/test/perf_profile.go:89]: Fixed

File: profile/gperf.go (hunk 15 above)
Comment (NiuJ1ao) [reply profile/gperf.go:16]: Removed

File: profile/gperf.go (hunk 14 above)
Comment (NiuJ1ao) [reply profile/gperf.go:89]: Done

File: profile/test/perf_profile.go [hunk 29]
```diff
@@ -0,0 +1,168 @@
...
+	events := strings.Split(*perfEvents, ",")
+	for _, e := range events {
+		titles = append(titles, e)
+	}
+
+	writer.Write(titles)
+	writer.Flush()
+}
+
+func plotResult() {
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/test/perf_profile.go:108]: Done

File: perf_bench_test.go [hunk 30]
```diff
@@ -47,104 +57,144 @@ func TestBenchRequestPerSecond(t *testing.T) {
...
+		serveMetrics[realRPS]++
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply perf_bench_test.go:152]: Done

File: profile/gperf.go (hunk 14 above)
Comment (NiuJ1ao) [reply profile/gperf.go:89]: I don't get why we need to run perf repeatedly. I thought we need to profile the entire invocations (requests). My implementation is start profiling just after VMs are booted and stop it after all invocations return. If we run perf for multiple rounds, it will just profile partial CPU counters of the whole test. Moreover, I think it is also difficult for users to determine the running time of perf.

Comment (NiuJ1ao) [reply profile/gperf.go]: Done

File: profile/test/perf_profile.go [hunk 31]
```diff
@@ -0,0 +1,168 @@
...
+	cmd := exec.Command("sudo", "/bin/sh", "-c", commandString)
+
+	if err := cmd.Run(); err != nil {
+		log.Fatalf("Setup benchmark error: %v\n", err)
+	}
+}
+
+func runTestBenchRequestPerSecond(vmNum int) {
+	envPath := "PATH=" + os.Getenv("PATH")
+	commandString := fmt.Sprintf("sudo env %s go test -v -run TestBenchRequestPerSecond -args -vm %d -funcNames %s -perfEvents %s",
```
[the remark this answers is not in the document]
Comment (NiuJ1ao) [reply profile/test/perf_profile.go]: Done

**Comment by NiuJ1ao** [comment]
I moved the UI to perf_bench_test.go. After the test starts, it runs `perf stat` when all VMs are booted. The perf stats are reported every N milliseconds for X seconds (user-defined). The warm-up time is set to the moment when all VMs are served at least once. The teardown time is when all invocations are returned. Then, the parser retrieves the middle part and calculate the average counters.

File: perf_bench_test.go [hunk 32]
```diff
@@ -24,127 +24,240 @@ package main
...
+	execTime := time.Since(tStart).Milliseconds()
+	atomic.AddInt64(avgExecTime, execTime)
+	log.Debugf("VM %s: returned in %d milliseconds", vmIDString, execTime)
+}
+
+// Returns a list of image names
+func getImages(t *testing.T) []string {
+	var (
+		images = map[string]string{
+			"helloworld":   "ustiugov/helloworld:var_workload",
```
Comment (ustiugov) [inline perf_bench_test.go:194]: use `vhiveease/helloworld:var_workload` instead of `ustiugov/helloworld:var_workload` for all functions

File: perf_bench_test.go [hunk 33]
```diff
@@ -24,127 +24,240 @@ package main
...
-		"helloworld": "ustiugov/helloworld:var_workload",
-		// "chameleon":    "ustiugov/chameleon:var_workload",
-		// "pyaes":        "ustiugov/pyaes:var_workload",
-		// "image_rotate": "ustiugov/image_rotate:var_workload",
-		// "json_serdes":  "ustiugov/json_serdes:var_workload",
-		// "lr_serving":   "ustiugov/lr_serving:var_workload",
-		// "cnn_serving":  "ustiugov/cnn_serving:var_workload",
-		// "rnn_serving":  "ustiugov/rnn_serving:var_workload",
-		// "lr_training":  "ustiugov/lr_training:var_workload",
+func getCPUIntenseRPS() map[string]int {
```
Comment (ustiugov) [inline perf_bench_test.go:252]: please add a description of this function in a comment

File: profile/perf_test.go [hunk 34]
```diff
@@ -0,0 +1,122 @@
...
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+type testCase struct {
+	warmTime, tearDownTime float64
+	expected               []map[string]float64
+}
+
```
Comment (ustiugov) [inline profile/perf_test.go:16]: please add a test for testing PerfStat.Run() function

File: rps_bench_profile.sh [hunk 35]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
Comment (ustiugov) [inline rps_bench_profile.sh:1]: this file can be removed

**Review by ustiugov — CHANGES_REQUESTED** [review]
This is very good progress! good job. Need to add a couple more tests and comments but looks good. 

In addition to my other comments, please add testing of the profile module to github actions. They should run on cloud-hosted runners (not on the self-hosted ones)

File: perf_bench_test.go (hunk 32 above)
Comment (NiuJ1ao) [reply perf_bench_test.go:194]: Done

File: perf_bench_test.go (hunk 33 above)
Comment (NiuJ1ao) [reply perf_bench_test.go:252]: Done

File: rps_bench_profile.sh (hunk 35 above)
Comment (NiuJ1ao) [reply rps_bench_profile.sh:1]: Done

File: profile/perf_test.go (hunk 34 above)
Comment (NiuJ1ao) [reply profile/perf_test.go:16]: Done
